pkg/controller/decorator/api/common/interface.go

```diff
@@ -0,0 +1,34 @@
+/*
+ *
+ * Copyright 2022. Metacontroller authors.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * https://www.apache.org/licenses/LICENSE-2.0
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * /
+ */
+
+package common
+
+import (
+	"metacontroller/pkg/apis/metacontroller/v1alpha1"
+	"metacontroller/pkg/controller/common/api"
+	commonv1 "metacontroller/pkg/controller/common/api/v1"
+
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+type WebhookRequestBuilder interface {
+	WithController(controller *v1alpha1.DecoratorController) WebhookRequestBuilder
```

Comment:
Same comment.
Have we have thought having following methods against a struct & hence avoid interface completely
- WithDecoratorController(...)
- WithCompositeController(...)

Reply:
It is not only about decorator/composite controller, there will be underlying implementation for `v1`/`v2` which will be different in the way how they construct the resource, and it is the main reason of introducing the interface.

Also, having `WithDecoratorController` and `WithCompositeController` bring complexity to the `build` method which I want to avoid